storage/bootstrap/bootstrapper/commitlog/source_prop_test.go

```diff
@@ -0,0 +1,235 @@
+			}
+
+			err = verifyShardResultsAreCorrect(values, result.ShardResults(), bootstrapOpts)
+			if err != nil {
+				return false, err
+			}
+			return true, nil
+		},
+		genPropTestInputs(testNamespaceID.String()),
+	))
+
+	props.TestingRun(t)
+}
+
+type propTestInput struct {
+	currentTime time.Time
+	writes      []generatedWrite
+}
+
+type generatedWrite struct {
+	series     commitlog.Series
+	datapoint  ts.Datapoint
+	unit       xtime.Unit
+	annotation ts.Annotation
+}
+
+func (w generatedWrite) String() string {
+	return fmt.Sprintf("ID = %v, Datapoint = %+v", w.series.ID.String(), w.datapoint)
+}
+
+func genPropTestInputs(ns string) gopter.Gen {
+	curriedGenPropTestInput := func(input interface{}) gopter.Gen {
+		inputs := input.([]interface{})
+		start := inputs[0].(time.Time)
+		numDatapoints := inputs[1].(int)
+		return genPropTestInput(start, numDatapoints, ns)
+	}
+	return gopter.CombineGens(
+		// Runs iterations of the test starting 1000 hours in the past/future
+		gen.TimeRange(time.Now().Add(-1000*time.Hour), 1000*time.Hour),
+		// Run iterations of the test with between 0 and 1000 datapoints
+		gen.IntRange(0, 1000),
+	).FlatMap(curriedGenPropTestInput, reflect.TypeOf(propTestInput{}))
+}
+
+func genPropTestInput(start time.Time, numDatapoints int, ns string) gopter.Gen {
+	return gen.SliceOfN(numDatapoints, genWrite(start, ns)).
+		Map(func(val interface{}) propTestInput {
+			return propTestInput{
+				currentTime: start,
+				writes:      val.([]generatedWrite),
+			}
+		})
+}
+
+func genWrite(start time.Time, ns string) gopter.Gen {
+	return gopter.CombineGens(
+		gen.Identifier(),
+		gen.TimeRange(start, 15*time.Minute),
+		// M3TSZ is lossy, so we want to avoid very large numbers with high amounts of precision
+		gen.Float64Range(-9999999, 99999999),
+		// Some of the commitlog bootstrapping code is O(N) with respect to the
+		// number of shards, so we cap it to prevent timeouts
+		gen.UInt32Range(0, 8192),
+	).Map(func(val []interface{}) generatedWrite {
+		id := val[0].(string)
+		t := val[1].(time.Time)
+		v := val[2].(float64)
+
+		return generatedWrite{
+			series: commitlog.Series{
+				ID:          ts.StringID(id),
+				Namespace:   ts.StringID(ns),
+				Shard:       idToShard(id),
+				UniqueIndex: idToIdx(id),
+			},
+			datapoint: ts.Datapoint{
+				Timestamp: t,
+				Value:     v,
+			},
+			unit: xtime.Nanosecond,
+		}
+	})
+}
+
+type globalMetricIdx struct {
+	sync.Mutex
+
+	idx     uint64
+	idToIdx map[string]uint64
+}
+
+type globalMetricShard struct {
+	sync.Mutex
+
+	shard     uint32
+	idToShard map[string]uint32
+}
+
+var metricIdx = globalMetricIdx{
+	idToIdx: make(map[string]uint64),
+}
+
+var metricShard = globalMetricShard{
+	idToShard: make(map[string]uint32),
+}
+
+// idToIDX ensures that each string series ID maps to exactly one UniqueIndex
+func idToIdx(s string) uint64 {
```

Comment:
nit: rename to something like `func seriesUniqueIndex(series string) uint64`

Reply:
done